pkgs/os-specific/darwin/apple-source-releases/xnu/default.nix

```
@@ -72,15 +72,19 @@ appleDerivation' (if headersOnly then stdenvNoCC else stdenv) (
   HOST_FLEX = "flex";
   HOST_BISON = "bison";
   HOST_GM4 = "m4";
-  MIGCC = "cc";
+  # use unwrapped clang to generate headers because wrapper is not compatible with a 32 bit -arch.
+  # aarch64 should likely do this as well and remove the --replace MACHINE_ARCH above
+  MIGCC = if stdenv.isx86_64 && lib.versionAtLeast stdenv.hostPlatform.darwinSdkVersion "11"
+    then "${lib.getBin buildPackages.stdenv.cc.cc}/bin/clang"
```

`mig` needs to target the build platform because it’s run at build time to generate headers.

```suggestion
    then "${lib.getBin pkgsBuildBuild.stdenv.cc.cc}/bin/clang"
```

---

I’d also add this can be done unconditionally. I do something similar in https://github.com/NixOS/nixpkgs/pull/256590/commits/37a87d0d1a5e9dbd1499e744bf68283da48eddb7, but I think this way is probably cleaner. I can’t remember why I didn’t specify a full path to for `MIGCC` in that commit.

---

i don't have aarch64 to test on. i suspect that making it unconditional means that MACHINE_ARCH doesn't need to get patched for aarch64 as well, and thus heads the comment saying that the header build needs to use 32 bit arch. it does change the layouts for the x64 darwin structures so i assume that the aarch64 ones have mismatched structure sizes as well.

i'd like this to be part of a followup given the lack of hardware and this would trigger the commit going into staging.